pkg/volume/aws_ebs/attacher.go

```diff
@@ -0,0 +1,236 @@
+/*
+Copyright 2016 The Kubernetes Authors All rights reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package aws_ebs
+
+import (
+	"fmt"
+	"os"
+	"path"
+	"strconv"
+	"time"
+
+	"github.com/golang/glog"
+	"k8s.io/kubernetes/pkg/api"
+	"k8s.io/kubernetes/pkg/util/exec"
+	"k8s.io/kubernetes/pkg/util/keymutex"
+	"k8s.io/kubernetes/pkg/util/mount"
+	"k8s.io/kubernetes/pkg/volume"
+)
+
+type awsElasticBlockStoreAttacher struct {
+	host volume.VolumeHost
+}
+
+var _ volume.Attacher = &awsElasticBlockStoreAttacher{}
+
+var _ volume.AttachableVolumePlugin = &awsElasticBlockStorePlugin{}
+
+// Singleton key mutex for keeping attach/detach operations for the same PD atomic
+var attachDetachMutex = keymutex.NewKeyMutex()
+
+func (plugin *awsElasticBlockStorePlugin) NewAttacher() (volume.Attacher, error) {
+	return &awsElasticBlockStoreAttacher{host: plugin.host}, nil
+}
+
+func (attacher *awsElasticBlockStoreAttacher) Attach(spec *volume.Spec, hostName string) error {
+	volumeSource, readOnly := getVolumeSource(spec)
+	VolumeID := volumeSource.VolumeID
+
+	// Block execution until any pending detach operations for this PD have completed
+	attachDetachMutex.LockKey(VolumeID)
+	defer attachDetachMutex.UnlockKey(VolumeID)
+
+	awsCloud, err := getCloudProvider(attacher.host.GetCloudProvider())
+	if err != nil {
+		return err
+	}
+
+	for numRetries := 0; numRetries < maxRetries; numRetries++ {
+		if numRetries > 0 {
+			glog.Warningf("Retrying attach for AWS PD %q (retry count=%v).", VolumeID, numRetries)
+		}
+
+		if _, err = awsCloud.AttachDisk(VolumeID, hostName, readOnly); err != nil {
+			glog.Errorf("Error attaching PD %q: %+v", VolumeID, err)
+			time.Sleep(errorSleepDuration)
+			continue
+		}
+		return nil
+	}
+
+	return err
+}
+
+func (attacher *awsElasticBlockStoreAttacher) WaitForAttach(spec *volume.Spec, timeout time.Duration) (string, error) {
+	awsCloud, err := getCloudProvider(attacher.host.GetCloudProvider())
+	if err != nil {
+		return "", err
+	}
+	volumeSource, _ := getVolumeSource(spec)
+	VolumeID := volumeSource.VolumeID
+	partition := ""
+	if volumeSource.Partition != 0 {
+		partition = strconv.Itoa(int(volumeSource.Partition))
+	}
+	devicePath, err := awsCloud.GetDiskPath(VolumeID)
```

Comment:
If I understand it correctly, `WaitForAttach` runs inside kubelet, right? If so, please add at least a TODO comment that we should not use cloudprovider and find the device ID from metadata server. Or implement the metadata server query :-).